### ISSUE TITLE
Report error on the @Mapping annotation when using invalid values

### ISSUE THREAD

**Issue by filiphr** [mapstruct/mapstruct#1383]
In `Mapping#fromMappingPrism` we do some validation checks and report an error on the method instead of the `@Mapping` annotation.

example:

```
@Mapper
public interface MyMapper {

    @Mapping( target = "id", expression = "java(String.valueOf( \"123\" )), constant = "1234") // error should be reported here
    Target map(Source source); // error report on this line
}
```

**Comment by MatteoJoliveau** [comment]
Hi @filiphr, I was looking to start contributing to MapStruct and stumbled upon this issue.
If you could explain a little what are the requirements I would love to tackle it and get a grasp on the project :)

**Comment by filiphr** [comment]
Hey @MatteoJoliveau it's really nice that you want to contribute to the project. We are always looking for contributors 😄.


Regarding this issue. The idea is to change the location of the reported error. Currently it is reported on the method, but it would be better if it is reported on the `@Mapping` annotation itself. The location for this changes are in [`Mappings#fromPrism`](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java#L99). The idea is in `message.printMessage(...)` to pass in the `AnnotationMirror` of the `MappingPrism` and the `AnnotationValue` from the appropriate location (`expression`, `defaultValue`, etc).

There are already tests that cover all those error messages:
* [`SourceConstantsTests#L115`](https://github.com/mapstruct/mapstruct/blob/master/processor/src/test/java/org/mapstruct/ap/test/source/constants/SourceConstantsTest.java#L115) -> In this test the first error should be on line 37 and the second one is OK.

You can find the others by searching for the error message in the code base.

If you need more help feel free to post more questions 😄. You can also find us in or gitter [mapstruct/mapstruct-users](https://gitter.im/mapstruct/mapstruct-users) chat. Usually we respond in the evenings there (CET time)

**Comment by MatteoJoliveau** [comment]
Hi @filiphr, thank you! I'll start digging in the codebase tomorrow and get in touch if needed :)

**Comment by dvfeinblum** [comment]
Hey there @filiphr.

I'm also looking to get my feet wet on Mapstruct and saw that this issue had quieted down. Thought I'd take a crack at it.

I'm running into something peculiar and was hoping you could give me a nudge. I went ahead and tried to fix just this one branch at [Mapping#L112](https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/Mapping.java#L112) by changing the if to:
```
if ( !mappingPrism.source().isEmpty() && mappingPrism.values.constant() != null ) {
    messager.printMessage(
           element,
           mappingPrism.mirror,
           mappingPrism.values.source(),
           Message.PROPERTYMAPPING_SOURCE_AND_CONSTANT_BOTH_DEFINED
    );
    return null;
}
```

However, regardless of which `value` I choose (`source`, `constant`, etc), I can't seem to get the line being reported in the message to change. Am I missing something silly here? Thanks!

**Comment by filiphr** [comment]
@dvfeinblum you only changed in the `Mapping` right? You didn't modify the test. I just tried with your example and that test fails with:

>org.junit.ComparisonFailure: 
>Expected :43L
>Actual   :37L

I would even say that you don't need the 3rd parameter. As both `source` and `constant` are wrong and we should only mark the entire annotation. From Java 9 if you put the `AnnotationValue` that particular element would be marked as an error (in Java 8 that parameter is not used).

**Comment by filiphr** [comment]
No problem. Looking forward to the PR 😀

**Comment by filiphr** [comment]
Thanks for the PR @dvfeinblum I have merged #1430
